### DISCUSSION TITLE
"Rebase and merge" should keep the original commits' timestamps

### DISCUSSION THREAD

**Discussion by andreynering** [community/community#110560] [Pull Requests]
### Select Topic Area

Product Feedback

### Body

When merging a PR by "rebase and merge", the commit date and time of every commit in the PR are overridden to the PR merge time.

This was always unexpected to me, as this is not how the actual `git rebase` command works. In my opinion, every commit should preserve its original commit date and time.

This has no relation to the actual order of the commits. By the way Git works, the commits should still all be moved to the top of the branch, but with no changed to its timestamp.

<img width="348" alt="Screenshot 2024-02-29 at 17 53 26" src="https://github.com/community/community/assets/7011819/4fe53277-3af5-4f16-9c43-13d0f1fc1664">

---

An additional thought is that maybe even "squash and merge" should keep the last commit date and time as the squashed commit timestamp. A PR often has even a single commit, so why not preserve its original timestamp? I think the current behavior is specially unexpected to "rebase and merge", though.

**Comment by alixodendhal** [comment]
I ran into this same issue. I cannot understand why they made this *customization* to do what we ***don't*** want. 

[It's noted here in their docs.](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github#rebasing-and-merging-your-commits)
> The rebase and merge behavior on GitHub deviates slightly from git rebase. Rebase and merge on GitHub will always update the committer information and create new commit SHAs, whereas git rebase outside of GitHub does not change the committer information when the rebase happens on top of an ancestor commit.
---
I have a use case where I have `staging` and `main` branches. I want `staging` to ff or rebase it's history onto `main` keeping their history identical for obvious reasons. 

This is what I have to deal with:
![image](https://github.com/user-attachments/assets/07e1b4b7-4d1e-40b9-999b-75f4788d8e72)

I had to write a custom github action to forcibly rebase backwards so that staging gets this nonsense re-written history. 

You can see that I have a commit hash then it gets re-written on Githubs rebase merge and I have to update staging. I shouldn't have to update staging. It had the commit before.

**Reply by anton-at-baraka** [reply]
It does not seem GH is going to fix that. Could you share the GH action you implemented?

**Comment by alixodendhal** [comment]
https://gist.github.com/alixodendhal/b630d9ff7ef7eaa7ed06e7b9fc51d7cc

You need to set an access token secret in the Repo. Since you are forcibly re-writing staging you will need to force push. Which is less than ideal. I allowed only this specific access token to re-write the commit history.

**Comment by ironchicken** [comment]
`git` even supports this behaviour: you can `git rebase` with the flag `--committer-date-is-author-date`. It feels like it should be quite trivial for the Github "rebase and merge" implementation to apply that flag.
